### ISSUE TITLE
Hidden Desktop appearing out of nowhere

### ISSUE THREAD

**Issue by Clyybber** [baskerville/bspwm#680]
So i have my 5 defined desktops but there is an additional desktop which i havent defined anywhere and its name is just "Desktop".

Heres my bspwmrc file:
```
#! /bin/sh

bspc config remove_disabled_monitor true

EXTERNAL_OUTPUT="VGA1"
INTERNAL_OUTPUT="eDP1"

xrandr |grep $EXTERNAL_OUTPUT | grep " connected "
if [ $? -eq 0 ]; then
    xrandr --output $INTERNAL_OUTPUT --off --output $EXTERNAL_OUTPUT --auto
    bspc monitor $EXTERNAL_OUTPUT -d I II III IV V
else
    xrandr --output $INTERNAL_OUTPUT --auto --output $EXTERNAL_OUTPUT --off
    bspc monitor $INTERNAL_OUTPUT -d I II III IV V
fi


sxhkd &
#if [ "$(bspc query -M | wc -l)" -eq "2" ]; then
#       bspc monitor $(bspc query -M | sed -n 1p) -d I II III
#       bspc monitor $(bspc query -M | sed -n 2p) -d IV V
#else
#       bspc monitor -d I II III IV V
#fi

bspc config border_width         5
bspc config window_gap          12

bspc config split_ratio          0.50
bspc config borderless_monocle   true
bspc config gapless_monocle      false

/usr/lib/polkit-gnome/polkit-gnome-authentication-agent-1 &

compton -b

wal -t -i "$(< "${HOME}/.cache/wal/wal")"

source $HOME/.config/bspwm/colors.sh

polybar top &
```

**Comment by msteen** [comment]
The name "Desktop" is the default desktop name used when a desktop has no given name. This is just speculation, but could it be caused by a timing issue between `bspwm` knowing the monitors have changed and your call to `bspc`? Am I correct to assume you problem only occurs when you have an external monitor connected? Or can a monitor be known to the system while being disconnected other than the builtin monitor? Anyway, this is likely what is happening:

1. Each monitor is assigned a desktop, which defaults to the name "Desktop". So now both the external and internal monitor have a desktop called "Desktop".
2. You then invoke `xrandr` this disables one of the monitors, but this is not known yet to `bspwm`, otherwise your --reset-desktops (-d) call would have removed the "Desktop" monitor.
3. You replace the desktops on the external monitor with those you have specified, so the internal will still have "Desktop", but the external now will have its "Desktop" desktop replaced with those you specified.
4. Then `bspwm` is made aware of the changes made in the monitor setup, so it will remove the monitor, causing the default "Desktop" desktop to be moved to the external monitor to prevent loss.

How to fix? Well you could use: `bspc subscribe monitor_rename` and then call `bspc monitor -d I II III IV V`. I do not specify which monitor, because I assumed you have only one monitor active at that point.

**Comment by baskerville** [comment]
@msteen Very good analysis. One thing I could add is an option to `subscribe`, for example `-c COUNT` that would create a subscriber that gets removed after having received exactly`COUNT` events.

**Comment by msteen** [comment]
@baskerville That would be great. I was looking for something like that and tried to experiment with how I could achieve this in bash in the general case (quit after receiving something N times from some piped producer), but have yet to find good solution. However it would probably be better if `bspc` would just provide this, because such a solution likely will be quite involved or shell specific.

**Comment by baskerville** [comment]
@msteen c744b74

*Note to myself*: try to take advantage of this in `tests/prelude`.

**Comment by msteen** [comment]
@baskerville Nice, but one suggestion I would make is to make sure `bspc query` closes when the output buffer closes. Right now it stays alive for the next event until it realizes it cannot write to the output and exists.

    bspc subscribe node_focus | { read -n 1; echo 5 }

Requires me to focus two windows before it exists. The first will cause `read -n 1` to complete and it will `echo 5`, but it requires a subsequent event before `bspc subscribe node_focus` closes as well.

**Comment by Clyybber** [comment]
Thank you! Since you are right, im only using 1 Monitor at a time your solution works perfectly.

**Comment by msteen** [comment]
@Clyybber That is because `bspc subscribe` is a long running process, which you generally want to [put in the background](https://bash.cyberciti.biz/guide/Putting_jobs_in_background). The whole point of `bspc subscribe` is to keep running and generate output when what you subscribed occured. And you would want to start it before you calls to `xrandr`, so you would do something like this:

    EXTERNAL_OUTPUT="VGA1"
    INTERNAL_OUTPUT="eDP1"

    { bspc subscribe monitor_remove & echo $!; wait $! } | stdbuf -oL -eL head -2 | {
      read pid
      read -n 1
      bspc monitor -d I II III IV V
      kill $pid
    } &

    xrandr | grep $EXTERNAL_OUTPUT | grep " connected "
    if [ $? -eq 0 ]; then
      xrandr --output $INTERNAL_OUTPUT --off --output $EXTERNAL_OUTPUT --auto
    else
      xrandr --output $INTERNAL_OUTPUT --auto --output $EXTERNAL_OUTPUT --off
    fi

**Comment by msteen** [comment]
@Clyybber I can confirm, that code I gave you should work, however, you made a typo in your `bspwmrc` and were using the wrong setting. It is `bspc config remove_disabled_monitors true` (monitors, not monitor), but more importantly it seems you have to use `bspc config remove_unplugged_monitors true` instead.

@baskerville Is this a bug? Should setting `remove_disabled_monitors` to true not have worked, otherwise I find the naming confusing. I would consider running `xrandr --output <monitor> --off` to be disabling your monitor and unplugging when you physically remove the connection with your monitor, right?

@baskerville Regarding closing `bspc subscribe` when the output in which it is piped is closed, should I create a separate issue for this? That would make by read and kill pid bits in my above examples obsolete.

**Comment by baskerville** [comment]
> I would consider running `xrandr --output <monitor> --off` to be disabling your monitor and unplugging when you physically remove the connection with your monitor, right?

That's correct, but unfortunately I have no way of knowing if any the `*_monitors` settings work as intended (I only have one monitor).

I should probably handle `xcb_randr_notify_event_t` events instead of querying for the monitor setup each time *something happens* (cf. *awesome WM*).

> Regarding closing `bspc subscribe` when the output in which it is piped is closed, should I create a separate issue for this?

Please go ahead.

**Comment by msteen** [comment]
@baskerville No easy way at least, surely you could ask some of your friends or family whether you could use their monitor to test it, but I can understand you prefer not to. I have three monitors (one decent and two small old flatscreens), so if you want I could test it for you. I just tested with:

    bspc config remove_unplugged_monitors true
    bspc config remove_disabled_monitors true
    bspc subscribe monitor_remove

And got no output after physically disconnecting my monitor. Disabling a monitor with `xrandr` will only work with both settings set to true, I just tested it with all combinations and that was the only one in which I got to see any output.

I made the issue regarding closing `bspc subscribe` when the output in which it is piped is closed: #681

**Comment by ottidmes** [comment]
@baskerville I have looked at the external monitor support and it seems to be correct, although I think the wording could be improved, otherwise the users intuition will not match the implementation.

My intuition was, as it is worded right now, the following:

`remove_disabled_monitors` meant that if I disable my monitor e.g. via `xrandr --output <output_name> --off` the the monitor with `<output_name>` would be removed as a monitor known to `bspwm`.
`remove_unplugged_monitors` meant that if you disconnect your monitor physically (unplug it) it would be removed as a monitor known to `bspwm`.

The actual meaning is the following:

`remove_disabled_monitors` means that if I disable my monitor e.g. via `xrandr --output <output_name> --off` the the monitor with `<output_name>` would be considered unplugged (disconnected).
`remove_unplugged_monitors` means that if your monitor is found to be unplugged (physically unplugged or with `remove_disabled_monitors` set to `true` also if disabled) it would be removed as a monitor known to `bspwm`.

I also looked whether it was possible to implement my intuition of `remove_unplugged_monitors`, and it is, but it is not as nice as I would have liked. I checked the code you linked, that of Awesome WM, but it won't generate an event when you disconnect your monitor. The only general way to implement it, i.e. not graphic card / driver dependent, seems to be to poll randr continuously. However you could make the poll interval high enough or configurable and then it probably would not be that bad.

**Comment by msteen** [comment]
@baskerville I have looked at the external monitor support and it seems to be correct, although I think the wording could be improved, otherwise the users intuition will not match the implementation.

My intuition was, as it is worded right now, the following:

`remove_disabled_monitors` meant that if I disable my monitor e.g. via `xrandr --output <output_name> --off` the the monitor with `<output_name>` would be removed as a monitor known to `bspwm`.
`remove_unplugged_monitors` meant that if you disconnect your monitor physically (unplug it) it would be removed as a monitor known to `bspwm`.

The actual meaning is the following:

`remove_disabled_monitors` means that if I disable my monitor e.g. via `xrandr --output <output_name> --off` the the monitor with `<output_name>` would be considered unplugged (disconnected).
`remove_unplugged_monitors` means that if your monitor is found to be unplugged (physically unplugged or with `remove_disabled_monitors` set to `true` also if disabled) it would be removed as a monitor known to `bspwm`.

I also looked whether it was possible to implement my intuition of `remove_unplugged_monitors`, and it is, but it is not as nice as I would have liked. I checked the code you linked, that of Awesome WM, but it won't generate an event when you disconnect your monitor. The only general way to implement it, i.e. not graphic card / driver dependent, seems to be to poll randr continuously. However you could make the poll interval high enough or configurable and then it probably would not be that bad.

**Comment by baskerville** [comment]
Thanks, I'll try to have sensible defaults regarding those settings.
